src/compiler/checker.ts

```diff
@@ -5041,10 +5041,24 @@ namespace ts {
                 if (relation === identityRelation) {
                     return signaturesIdenticalTo(source, target, kind);
                 }
-                if (target === anyFunctionType || source === anyFunctionType) {
+
+                // In type argument inference, we need to take a common supertype of multiple candidates. It is possible
+                // that one of these candidates is anyFunctionType, and the other is a function type with actual signatures.
+                // Intuitively, the one with signatures is a better inference. We encode that here by saying that a type
+                // with signatures is never a subtype of anyFunctionType (but it is assignable). This makes the type with
+                // signatures win over anyFunctionType.
+                if (source === anyFunctionType || (relation === assignableRelation && target === anyFunctionType)) {
                     return Ternary.True;
                 }
+
                 let sourceSignatures = getSignaturesOfType(source, kind);
+                if (target === anyFunctionType) {
+                    Debug.assert(relation === subtypeRelation);
+                    if (sourceSignatures.length > 0) {

```

Comment:
Ah, basically, if the source has no signatures, I think it's still okay for it to be a subtype of anyFunctionType. There is nothing that necessarily makes it better than anyFunctionType. The case I care about is when it *does* have signatures.